jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/CachedJMSConnectionFactory.java

```diff
@@ -0,0 +1,277 @@
+            return null;
+        }
+        try {
+            connection.start();
+        } catch (JMSException e) {
+            if (cachedConnection != null) {
+                resetCache();
+                getConnection(userName, password);
+            } else {
+                logger.error(
+                        "JMS Exception while starting connection for factory '" + this.connectionFactoryString + "' " +
+                        e.getMessage());
+                throw e;
+            }
+
+        }
+        return connection;
+    }
+
+    @Override public Connection createConnection(String userName, String password) throws JMSException {
+        Connection connection;
+        if (userName == null || password == null) {
+            connection = super.createConnection();
+        } else {
+            connection = super.createConnection(userName, password);
+        }
+        if (this.cacheLevel >= JMSConstants.CACHE_CONNECTION) {
+            cachedConnection = connection;
+        }
+        return connection;
+    }
+
+    @Override public Session getSession(Connection connection) throws JMSConnectorException {
+        if (cachedSession == null) {
+            return createSession(connection);
+        } else {
+            return cachedSession;
+        }
+    }
+
+    @Override public Session createSession(Connection connection) throws JMSConnectorException {
+        Session session = super.createSession(connection);
+        if (this.cacheLevel >= JMSConstants.CACHE_SESSION) {
+            cachedSession = session;
+        }
+        return session;
+    }
+
+    @Override public MessageConsumer getMessageConsumer(Session session, Destination destination)
+            throws JMSConnectorException {
+        MessageConsumer messageConsumer = null;
+        if (cachedMessageConsumer == null) {
+            messageConsumer = createMessageConsumer(session, destination);
+        } else {
+            messageConsumer = cachedMessageConsumer;
+        }
+        return messageConsumer;
+    }
+
+    @Override public MessageConsumer createMessageConsumer(Session session, Destination destination)
+            throws JMSConnectorException {
+        MessageConsumer messageConsumer = super.createMessageConsumer(session, destination);
+        if (this.cacheLevel >= JMSConstants.CACHE_CONSUMER) {
+            cachedMessageConsumer = messageConsumer;
+        }
+        return messageConsumer;
+    }
+
+    @Override public MessageProducer getMessageProducer(Session session, Destination destination)
+            throws JMSConnectorException {
+        MessageProducer messageProducer = null;
+        if (cachedMessageConsumer == null) {
+            messageProducer = createMessageProducer(session, destination);
+        } else {
+            messageProducer = cachedMessageProducer;
+        }
+        return messageProducer;
+    }
+
+    @Override public MessageProducer createMessageProducer(Session session, Destination destination)
```

Comment:
To me this is like a wrapper. Extended functionalities should be documented. Otherwise need to have doc inherit annotation.